### ISSUE TITLE
Best practice for training YOLOv8 with very small datasets

### ISSUE THREAD

**Issue by haimat** [ultralytics/ultralytics#6201]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I understand that it is always preferrable to have many hundrets or even better thousands of examples per class in the training dataset of a YOLOv8 model. However, in some projects this is not an option - for example for a very specific problem it might simply not be possible to generate so many samples of the object to detect.

So let's say we have only a very few images of the object we want to detect, something between 10 and 30. I understand that one way to tackle this problem is to create more example images by augmenting the few original images we have. But even with that approach we can only generate so many new images - in the end we still might not end up with thousands of training images.

So generally speaking, what is the "best practice" approach for training a YOLOv8 model when we only have such a small training dataset?

**Comment by PiyushKamboj1503** [comment]
@haimat best practice here would be to use the .pt pretrained models. Also, if your class is in the already trained classes, you can use the weight transfer

**Comment by glenn-jocher** [comment]
Hello @haimat,

Dealing with a very small dataset for training an object detection model is indeed challenging, but there are strategies you can employ to increase the chances of achieving good performance with YOLOv8:

1. **Pretrained Models**: As you've mentioned, starting with a pre-trained model is important. These models have learned features from a vast and diverse dataset, which can be beneficial even when your target classes are not part of the pretrained dataset.

2. **Transfer Learning**: Even if the pre-trained classes do not overlap with your industrial-specific objects, the learned features can still be useful. You can fine-tune a pretrained YOLOv8 model on your small dataset, which allows it to adapt the general features it has learned to your specific task.

3. **Data Augmentation**: This is crucial when dealing with limited data. Techniques like random rotations, flipping, scaling, cropping, and color space transformations can generate more varied training examples. YOLOv8 has built-in data augmentation that you can leverage during training.

4. **Overfitting Prevention**: Since your dataset is small, your model might overfit quickly. To prevent this, use strong regularization techniques, keep the network's capacity in check (i.e., avoid using a very large model), and closely monitor your validation performance.

5. **Transfer from Similar Classes**: If the pretrained model contains classes that are visually or contextually similar to your target objects, you may have better transfer learning results, as the model has already learned relevant features.

6. **Few-Shot Learning**: Look into few-shot learning techniques that are designed to work well with a limited number of training samples. These methods are often more robust to overfitting when dealing with small datasets.

7. **Ensemble Methods**: If possible, train several models with different initializations or architectures, and then combine their predictions. This can sometimes yield better performance than any single model.

8. **Domain-Specific Pretraining**: If you have access to a larger, yet still relevant dataset from a similar domain, pretraining on this before fine-tuning on your small dataset can be beneficial.

9. **Hyperparameter Tuning**: Be prepared to experiment with different hyperparameters, particularly those controlling the model's capacity to learn and generalize.

10. **Synthetic Data**: If applicable, consider generating synthetic data that closely resembles your actual objects of interest to supplement your training dataset.

11. **

**Comment by haimat** [comment]
@glenn-jocher Thank you very much for your reply, Glenn.

**Comment by glenn-jocher** [comment]
@haimat hello,

Training YOLOv8 with a very small dataset is a common challenge, but there are strategies to improve performance:

1. **Use Pretrained Weights**: Start with the weights of a pretrained YOLOv8 model as the foundation for your training. This leverages the knowledge gained from large datasets and often leads to better performance even when the exact classes do not match your specialized use case.

2. **Fine-tuning**: Fine-tune the pretrained model on your small dataset. This allows the model to adjust the learned features to better match your specific target objects.

3. **Data Augmentation**: Make the most of your data through augmentation strategies inbuilt into YOLOv8, such as random cropping, flipping, and color adjustments. Being creative here can substantially increase the effective size of your dataset without collecting new images.

4. **Regularization**: Use regularization techniques to prevent overfitting, such as dropout or weight decay. Carefully monitor the training to ensure the model is generalizing well and not just memorizing your limited training data.

5. **Few-shot Learning**: Consider methods specially designed for training on limited data, which might help the model to learn from a small number of examples more effectively.

6. **Hyperparameter Optimization**: Optimize the hyperparameters thoughtfully. This may include adjusting the learning rate, batch size, or number of epochs to find the best settings for your dataset.

7. **Model Size Selection**: With a small dataset, a smaller model can sometimes generalize better. Avoid models that are too large as they may be more prone to overfitting.

8. **Gradual Unfreezing**: Gradually unfreeze your model layers—from top to bottom—allowing for selective fine-tuning.

Lastly, while strategies like synthetic data generation and ensembling may be beyond the scope of YOLOv8's built-in tools and are not covered in this response, they can also be considered in larger pipeline development if resources allow for it.

Let's keep the dialogue open, and feel free to reach out if you face specific issues during your training process.

Best,
Glenn Jocher

**Comment by haimat** [comment]
@glenn-jocher Thanks for your feedback. Two questions:
 - Do you have more info on few-shot learning (in the context of YOLOv8)?
 - Is the more documentation or howtos on how to (un-)freeze layers in YOLOv8?

**Comment by glenn-jocher** [comment]
@haimat

Few-shot learning methods aim to train models that can learn from a small number of examples, making them well-suited for scenarios with limited training data. In the context of YOLOv8, few-shot learning can be implemented by utilizing techniques such as meta-learning, siamese networks, or prototype-based methods. At this time, there may not be built-in few-shot learning methods specifically tailored for YOLOv8, but it is an active area of research and you may find custom implementations or adaptations from related object detection frameworks.

Regarding freezing and unfreezing layers in YOLOv8, detailed documentation and how-tos specific to YOLOv8 can be found within the Ultralytics documentation and educational resources. The process typically involves freezing the weights of certain layers during initial training, then selectively unfreezing and fine-tuning them in subsequent training stages.

For further guidance on YOLOv8 and its implementation details, it is recommended to refer to the official documentation, code comments, and community forums such as GitHub discussions, where contributors and users regularly share insights and best practices.

If you encounter specific challenges or have further questions, please feel free to reach out for assistance.

Best regards,
Glenn Jocher

**Comment by glenn-jocher** [comment]
@haimat,

Few-shot learning involves training a model on a small number of examples per class, making it well-suited for scenarios with limited data. Utilizing this approach with YOLOv8 would typically involve techniques such as meta-learning, transfer learning, or model parameter adaptation to improve the model's ability to generalize from a small sample size. While few-shot learning is not directly integrated into YOLOv8, some research around integrating such techniques into the YOLOv8 framework may offer insights into using few-shot learning within the context of YOLOv8.

Regarding freezing and unfreezing layers in YOLOv8, the model's architecture and the training process are primarily handled within the YOLOv8 implementation and training script provided by the Ultralytics team. The YOLOv8 training script includes functionality for freezing and unfreezing layers during the training process, allowing for more nuanced model training. The specific approach to freezing and unfreezing layers, along with corresponding documentation or tutorials, can be found within the YOLOv8 repository, typically within the training script and associated documentation. Additionally, the Ultralytics team is actively maintaining the repository and can provide guidance on effectively freezing and unfreezing layers within YOLOv8 to suit your specific training needs.

For more detailed information on few-shot learning and layer freezing/unfreezing within the context of YOLOv8, I recommend referencing the YOLOv8 repository and its associated documentation. Furthermore, engaging with the Ultralytics team and the YOLO community can provide valuable insights and guidance as you navigate these specific aspects of training with YOLOv8.

Warm regards,
Glenn Jocher

**Comment by glenn-jocher** [comment]
@haimat You're welcome! Few-shot learning involves training a model with a minimal number of examples per class. In the context of YOLOv8, few-shot learning techniques usually involve modifications to the loss functions or network architectures to effectively utilize the limited training samples. While YOLOv8 doesn't have specific support for few-shot learning out-of-the-box, you can explore research papers and implementations that focus on few-shot object detection to adapt techniques to suit your needs.

As for (un-)freezing layers in YOLOv8, you can utilize the `--weights`, `--cfg`, and `--resume` flags along with the `--epochs` flag during training to control the freezing and unfreezing of layers. Additionally, a commonly used practice is to initially freeze all layers except the final few, and gradually unfreeze additional layers as training progresses. While there might not be explicit how-tos for this in the YOLOv8 documentation, the community and YOLOv8 GitHub issues may have discussions and examples that can guide you in this respect.

Feel free to ask any more questions or seek further clarification, and I'm here to assist throughout your training journey.

Best, Glenn Jocher

**Comment by glenn-jocher** [comment]
@haimat,

Few-shot learning aims to train a model using a minimal amount of labeled data per class. In the context of YOLOv8, while there isn’t a native few-shot learning setup, you can adapt your training pipeline to support this approach.

Here are a few common methods for few-shot learning in object detection:

1. **Meta-Learning**: Meta-learning algorithms, like MAML (Model-Agnostic Meta-Learning), aim to train a network in a way that allows it to adapt quickly to new tasks with minimal training. This can be applied to object detection tasks although the integration with YOLOv8 would require custom implementation.

2. **Data Augmentation**: Use aggressive data augmentation strategies, especially when training on a small number of samples per class. Techniques like mixup, cutmix, or semantic augmentation can help the model generalize better from few examples.

3. **Pretraining on Similar Classes**: If available, pretrained models or features on related classes can provide a good initialization when fine-tuning your YOLOv8 model on the few-shot dataset.

Regarding (un-)freezing layers in YOLOv8, here's an outline of the typical process:

1. **Freezing Layers**: When performing transfer learning or initial training, you might choose to freeze layers by setting their `requires_grad` attribute to `False`. In YOLOv8, the lower convolutional layers are good candidates for freezing during initial training.

2. **Unfreezing for Fine-Tuning**: During fine-tuning, you can gradually unfreeze layers starting from the top and moving downwards. This allows the model to fine-tune its parameters on the specific dataset while still retaining the general features learned from the pretrained weights.

For more specific documentation on implementation, I recommend referring to the official YOLOv8 repository and the associated documentation in its README. Additionally, exploring community-contributed tutorials or blog posts might provide insights into practical implementation details for few-shot learning and layer freezing in YOLOv8.

I hope this guidance provides a good starting point for your work. Don’t hesitate to reach out if you have further inquiries or need more detailed information on these topics.

Best regards,
Glenn Jocher

**Comment by haimat** [comment]
@glenn-jocher Thanks Glenn, that helps a lot :+1:

**Comment by glenn-jocher** [comment]
@haimat, you're welcome! I'm glad to hear the information was helpful. Good luck with your YOLOv8 training efforts, and should you have any more questions down the line, feel free to reach out again. Your successes and challenges contribute to the growth of the YOLO community and Ultralytics. Happy detecting!

**Comment by haimat** [comment]
@glenn-jocher Many thanks back to you, Glenn. It's really great that you share this awesome piece of software with the community. Keep up the good work :+1:

**Comment by glenn-jocher** [comment]
@haimat, it's truly a pleasure, and the positivity and contributions from users like you enhance the project for everyone involved. Remember, this is a collaborative effort and your feedback and experiences are invaluable. Enjoy working with YOLOv8!

**Comment by VoingShrikant** [comment]
Hey, @glenn-jocher I have a small dataset of around 250 images and I was trying to finetune the YOLOV8m model over that dataset. However, the finetuned model has higher losses. And I want to tweak the model parameters to improve the model performance. So, how to do that for the yolo8m, and which parameters are important in my case?

**Comment by H0128Z** [comment]
@haimat 
Hello, I'm also trying to apply MAML to YOLOv8 for few-shot learning. I was wondering if I could ask how you managed to resolve the issue mentioned above?

Thank you!

**Comment by haimat** [comment]
@H0128Z Hi, wie ended up augmenting the existing images we had ...

**Comment by H0128Z** [comment]
@haimat  oh, thanks for your reply.

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has been merged in PR #19870, which introduces the new CutMix image augmentation feature in Ultralytics 8.3.119. With CutMix, users can now apply an advanced augmentation technique that improves model robustness by simulating realistic occlusions and challenging scenarios during training. This update adds a cutmix option alongside existing methods like mosaic and mixup, updates configuration files, and enhances the documentation with detailed guides and visuals. The integration is seamless, allowing flexible tuning and easy adoption within the training pipeline.

Special thanks to the PR author and contributors for their efforts in bringing this feature to Ultralytics.

To test if this resolves your issue, please update your package with:
```
pip install -U ultralytics>=8.3.119
```
Alternatively, you can install the latest code from the main branch:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Feel free to wait for the next official PyPI release if you prefer.

Your feedback is appreciated—please let us know if this update resolves the problem or if you encounter any further issues. Thank you for reporting this and helping improve Ultralytics! 🙏
